sql,externalconn: add ALTER EXTERNAL CONNECTION implmentation

edwardguo-crl opened the pull request:
### Summary

 Add implementation for ALTER EXTERNAL CONNECTION

 Fixes #98610
 
 Release note (sql change): Users can now ALTER EXTERNAL CONNECTION to change the external connection URI when granted UPDATE privilege on EXTERNAL CONNECTION.

 
 ### Local testing
 
 ```shell

root@localhost:26257/defaultdb> SHOW EXTERNAL CONNECTIONS;
  connection_name |     connection_uri      | connection_type
------------------+-------------------------+------------------
  test            | nodelocal://1/alterguo  | STORAGE
  test1           | nodelocal://1/alterguo1 | STORAGE
(2 rows)

Time: 10ms total (execution 10ms / network 0ms)

root@localhost:26257/defaultdb> ALTER EXTERNAL CONNECTION IF EXISTS test1 AS 'nodelocal://1/alterguo2';
ALTER EXTERNAL CONNECTION

Time: 146ms total (execution 146ms / network 0ms)

root@localhost:26257/defaultdb> SHOW EXTERNAL CONNECTIONS;
  connection_name |     connection_uri      | connection_type
------------------+-------------------------+------------------
  test            | nodelocal://1/alterguo  | STORAGE
  test1           | nodelocal://1/alterguo2 | STORAGE
(2 rows)

Time: 3ms total (execution 3ms / network 1ms)

root@localhost:26257/defaultdb> ALTER EXTERNAL CONNECTION IF EXISTS test2 AS 'nodelocal://1/alterguo2';
ALTER EXTERNAL CONNECTION

Time: 2ms total (execution 2ms / network 0ms)

root@localhost:26257/defaultdb> SHOW EXTERNAL CONNECTIONS;
  connection_name |     connection_uri      | connection_type
------------------+-------------------------+------------------
  test            | nodelocal://1/alterguo  | STORAGE
  test1           | nodelocal://1/alterguo2 | STORAGE
(2 rows)

Time: 2ms total (execution 2ms / network 0ms)

root@localhost:26257/defaultdb> ALTER EXTERNAL CONNECTION test2 AS 'nodelocal://1/alterguo2';
ERROR: external connection with name test2 does not exist
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/149869)
<!-- Reviewable:end -->

kev-cao commented on pkg/sql/alter_external_connection.go:
```diff
@@ -0,0 +1,136 @@
+		return errors.Wrap(err, "failed to get owner ID for External Connection")
+	}
+
+	ownerID := tree.MustBeDOid(row[0]).Oid
+	ex.SetOwnerID(ownerID)
+
+	if err := ex.Update(params.ctx, txn); err != nil {
+		return errors.Wrap(err, "failed to alter external connection")
+	}
+
+	grantStatement := fmt.Sprintf(`GRANT ALL ON EXTERNAL CONNECTION "%s" TO %s`,
+		name, p.User().SQLIdentifier())
+	_, err = txn.ExecEx(
+		params.ctx, "grant-on-alter-external-connection", txn.KV(),
+		sessiondata.NodeUserSessionDataOverride, grantStatement)
```
Should someone that is updating an external connection now be considered the owner of that connection?

And should updating an external connection grant you all privileges for that connection automatically?

msbutler replied:
see comment above: this cmd should simply swap the external conn URI and no other side effects (e.g. transferring ownership)

kev-cao commented on pkg/cloud/externalconn/record.go:
```diff
@@ -354,6 +354,26 @@ func (e *MutableExternalConnection) Create(ctx context.Context, txn isql.Txn) er
+		setClauses = append(setClauses, fmt.Sprintf("%s = $%d", col, i+1))
+	}
+
+	updateQuery := fmt.Sprintf("UPDATE system.external_connections SET %s, updated = now() WHERE connection_name = '%s'",
+		strings.Join(setClauses, ", "), e.ConnectionName())
```
I think it just didn't catch it then. I'm seeing this:
<img width="825" height="101" alt="image" src="https://github.com/user-attachments/assets/3ab4f6f7-0ba3-4cc7-820f-0d8c9906254f" />

I think this would be more readable:
<img width="796" height="143" alt="image" src="https://github.com/user-attachments/assets/8f5d74ec-06d0-4535-aaff-98d7ba1fee4a" />

edwardguo-crl commented on pkg/sql/alter_external_connection.go:
```diff
@@ -0,0 +1,136 @@
+
+func (p *planner) alterExternalConnection(params runParams, n *tree.AlterExternalConnection) error {
+	txn := p.InternalSQLTxn()
+	if err := params.p.CheckPrivilege(params.ctx, syntheticprivilege.GlobalPrivilegeObject, privilege.EXTERNALCONNECTION); err != nil {
```
Right now, it's only checking the general external connection privilege? Should we also check specific privilege of this external connection object?

msbutler replied:
let's follow the same approach used in DROP EXTERNAL CONNECTION, but lets ensure the user has the UPDATE priv:
https://github.com/edwardguo-crl/cockroach/blob/feature-alter-ext-connection-impl/pkg/sql/drop_external_connection.go#L51

msbutler replied:
could you also add a test around privs once you update the logic here?

msbutler commented on pkg/sql/alter_external_connection.go:
```diff
@@ -0,0 +1,136 @@
+
+	ec, err := externalconn.LoadExternalConnection(params.ctx, name, txn)
+	if err != nil {
+		if ok := strings.Contains(err.Error(), "does not exist"); ok && n.IfExists {
```
instead of doing string comparison, you should check the explicit error type, i.e.:
```
errors.Is(err, externalConnectionNotFoundError) 
```

msbutler commented on pkg/sql/alter_external_connection.go:
```diff
@@ -0,0 +1,136 @@
+
+	var skipCheckingExternalStorageConnection bool
+	var skipCheckingKMSConnection bool
+	if tk := params.ExecCfg().ExternalConnectionTestingKnobs; tk != nil {
```
since we arent using these testing knobs in this code, i don't think we need to have them here

msbutler commented on pkg/sql/alter_external_connection.go:
```diff
@@ -0,0 +1,136 @@
+		return err
+	}
+
+	ex := externalconn.NewMutableExternalConnection()
```
could this logic get moved to after you create the external conn env?

msbutler commented on pkg/sql/alter_external_connection.go:
```diff
@@ -0,0 +1,136 @@
+		return err
+	}
+
+	ec, err := externalconn.LoadExternalConnection(params.ctx, name, txn)
```
nit: we try not to use abbreviation variables. variables should be self explanatory. call this `existingConn` and `ex` `newMutableConn`?

msbutler replied:
but also, could we clean this logic up by casting the existingConn to a `mutableExternalConnection`, and then swapping the URI? I don't think this function should be modifying the owner of the external conn.

msbutler reviewed (changes requested):
This PR should also have a release note.

edwardguo-crl commented on pkg/sql/alter_external_connection.go:
```diff
@@ -87,44 +77,24 @@ func (p *planner) alterExternalConnection(params runParams, n *tree.AlterExterna
 		params.ExecCfg().DistSQLSrv.ExternalStorageFromURI,
-		skipCheckingExternalStorageConnection,
-		skipCheckingKMSConnection,
+		false,
```
Since we are not using the testingKnobs, I set it here to False.

CC: @msbutler

msbutler reviewed (commented):
Nice! feel free to squash. Also, this needs a release note (in the commit and PR message)

edwardguo-crl commented on pkg/sql/logictest/testdata/logic_test/alter_external_connection:
```diff
@@ -0,0 +1,58 @@
+statement ok
```
No wonder there was so many tenant testing running to this change. :)

edwardguo-crl committed:
sql,externalconn: add ALTER EXTERNAL CONNECTION implmentation

Add implementation for ALTER EXTERNAL CONNECTION

Release note (sql change): Users can now ALTER EXTERNAL CONNECTION to
change the external connection URI when granted UPDATE privilege on
EXTERNAL CONNECTION.

Fixes #98610

edwardguo-crl commented:
bors r+
